gcp_variant_transforms/beam_io/vcf_header_io.py

```diff
@@ -0,0 +1,307 @@
+
+  def _merge_header_fields(self, source, to_merge):
+    """Modifies ``source`` to add any keys from ``to_merge`` not in ``source``.
+
+    Args:
+      source (dict): Source header fields.
+      to_merge (dict): Header fields to merge with ``source``.
+    Raises:
+      ValueError: If the header fields are incompatible (e.g. same key with
+        different types or numbers).
+    """
+    for key, to_merge_value in to_merge.iteritems():
+      if key not in source:
+        source[key] = to_merge_value
+      else:
+        source_value = source[key]
+        if source_value != to_merge_value:
+          raise ValueError(
+              'Incompatible number of types in header fields: %s, %s' % (
+                  source_value, to_merge_value))
+
+
+class _VcfHeaderSource(filebasedsource.FileBasedSource):
+  """A source for reading VCF file headers.
+
+  Parses VCF files (version 4) using PyVCF library.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64 kB
+
+  class ReadBuffer(object):
+    # A buffer that gives the buffered data and next position in the
+    # buffer that should be read.
+
+    def __init__(self, data, position):
+      self._data = data
+      self._position = position
+
+    @property
+    def data(self):
+      return self._data
+
+    @data.setter
+    def data(self, value):
+      assert isinstance(value, bytes)
+      self._data = value
+
+    @property
+    def position(self):
+      return self._position
+
+    @position.setter
+    def position(self, value):
+      assert isinstance(value, (int, long))
+      if value > len(self._data):
+        raise ValueError('Cannot set position to %d since it\'s larger than '
+                         'size of data %d.', value, len(self._data))
+      self._position = value
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfHeaderSource, self).__init__(file_pattern,
+                                           compression_type=compression_type,
+                                           validate=validate,
+                                           splittable=False)
+    self._coder = coders.StrUtf8Coder()
+    self._buffer_size = buffer_size
+
+  def read_records(self, file_name, range_tracker):
+    try:
+      vcf_reader = vcf.Reader(fsock=self._read_headers(file_name))
+    except StopIteration:
+      raise ValueError('%s has no header.', file_name)
+
+    yield VcfHeader(infos=vcf_reader.infos,
+                    filters=vcf_reader.filters,
+                    alts=vcf_reader.alts,
+                    formats=vcf_reader.formats,
+                    contigs=vcf_reader.contigs,
+                    samples=vcf_reader.samples)
+
+  def _read_headers(self, file_name):
+    read_buffer = _VcfHeaderSource.ReadBuffer('', 0)
```

Comment:
I wonder if we can just use the readline() functionality that Filesystem provides: https://github.com/googlegenomics/gcp-variant-transforms/blob/master/gcp_variant_transforms/libs/vcf_header_parser.py#L102

Reply:
Wow, this is an incredible improvement! The change cut out like 90% of the code in VcfHeaderSource!